soliditygen/generator.go

```diff
@@ -0,0 +1,159 @@
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type Definition struct {
+	Name string
+	Text string
+}
+
+// WriteSolidityFromURL fetches JSON from the given URL, infers Solidity structs,
+// and writes a .sol file named <contractName>Types.sol into the specified
+// contractDirectory. Returns the path to the created file or an error.
+func WriteSolidityFromURL(url, contractName, contractDirectory string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", fmt.Errorf("fetching JSON: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("non-OK status code: %d", resp.StatusCode)
+	}
+	defer resp.Body.Close()
+
+	bodyBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("reading response body: %w", err)
+	}
+
+	var data interface{}
+	if err := json.Unmarshal(bodyBytes, &data); err != nil {
+		return "", fmt.Errorf("unmarshaling JSON: %w", err)
+	}
+
+	definitions := convertToSolidityStruct(contractName, data)
+
+	contractName += "Types"
+
+	topDef := definitions[len(definitions)-1]
+
+	var solBuilder strings.Builder
+
+	solBuilder.WriteString("// SPDX-License-Identifier: GPL-3.0\n")
+	solBuilder.WriteString("// Code generated - DO NOT EDIT.\n")
+	solBuilder.WriteString("// This file is a generated and any manual changes will be lost.\n")
+	solBuilder.WriteString("pragma solidity >=0.8.25 <0.9.0;\n\n")
+	solBuilder.WriteString("/**\n")
+	solBuilder.WriteString(" * @title " + contractName + "\n")
+	solBuilder.WriteString(" * @notice Solidity contract that declares structs derived from JSON.\n")
+	solBuilder.WriteString(" */\n")
+
+	solBuilder.WriteString("contract " + contractName + " {\n\n")
+
+	for _, def := range definitions {
+		solBuilder.WriteString(def.Text)
+		solBuilder.WriteString("\n\n")
+	}
+
+	solBuilder.WriteString("function useAllTypes(")
+	solBuilder.WriteString(fmt.Sprintf("%s memory _%s", topDef.Name, strings.ToLower(topDef.Name)))
+	solBuilder.WriteString(") external {\n")
+	solBuilder.WriteString("    // This function doesn't do anything but ensures the top-level struct is referenced.\n")
+	solBuilder.WriteString("}\n")
+
+	solBuilder.WriteString("}\n")
+
+	solFileName := filepath.Join(contractDirectory, contractName+".sol")
+
+	if err := os.MkdirAll(contractDirectory, 0o755); err != nil {
+		return "", fmt.Errorf("creating directory: %w", err)
+	}
+	if err := os.WriteFile(solFileName, []byte(solBuilder.String()), 0o644); err != nil {
+		return "", fmt.Errorf("writing solidity file: %w", err)
+	}
+
+	return solFileName, nil
+}
+
+// convertToSolidityStruct recursively generates a child-first list of Solidity struct definitions.
+func convertToSolidityStruct(structName string, value interface{}) []Definition {
+	obj, ok := value.(map[string]interface{})
+	if !ok {
+		fallback := fmt.Sprintf("struct %s {\n    string fallback;\n}", structName)
+		return []Definition{{
+			Name: structName,
+			Text: fallback,
+		}}
+	}
```

Comment:
any reason for not doing this? from the code it looks like it's always called with a `map[string]interface{}` param

```suggestion
func convertToSolidityStruct(structName string, value map[string]interface{}) []Definition {
```